rfcs/037-index-query-composition.md

```diff
@@ -0,0 +1,115 @@
+        - user: an optional user that executed the query (should be Amundsen user ID)
+    - Relations:
+        - Table > Query
+        - User > Query
+2. Query Execution
+    - Attributes:
+        - query: the query model defined above
+        - start time: The start time of the window that the aggregation represents. Query executions can be dropped from the database when start time > X in order to update the most commonly used joins and wheres.
+        - execution_count: the total # of times the query was executed in the window
+        - window_duration: a string representation of the window length. Amundsen will not enforce non-overlapping windows, users should schedule a job to create query executions on defined cadence.
+    - Relations:
+        - Query >> Query Execution
+3. Join
+    - Attributes:
+        - left_table: Table Metadata for the left table in a join
+        - right_table: Table Metadata for the right table in a join
+        - left_column: Column Metadata for the left column in a join
+        - right_column: Column Metadata for the right column in a join
+        - join_type: a string representing the join type, could be `inner join`, `left outer join`, etc
+        - join_operator: a string representing the operator, could be `=`, `>`, `<=`, etc.
+        - join_sql: the sql code for this join (e.g. `table join table2 on table.a = table2.b`)
+        - query: The query model defined above
+    - Relations:
+        - Column >> Join
+        - Query >> Join
+4. Where
+    - Attributes:
+        - tables: a list of Table Metadata representing each table used in the query. The columns in the table metadata should be used in the where clause
+        - where_clause: a string representation of the SQL where (e.g. `field < 4` or `val is null`)
+        - query_metadata: The query model defined above
+    - Relations:
+        - Column >> Where
+        - Query >> Where
+
+## Drawbacks
+
+Parsing query logs can be a large effort, therefore this approach will be designed in a way such that implementers will be able to send a raw query into Amundsen without the other components (executions, joins and wheres) but will loose the ability for the features built on top of those data assets such as ordering by most common query, showing the most common filters, etc. However, Amundsen is not intended to hold a copy of each query each time it is executed and implementers should take some effort to dedupe queries.
+
+## Alternatives
+
+- Do not index queries and query components
+- Only index queries, not components of queries
+
+## Prior art
+
+[PR 27](https://github.com/amundsen-io/rfcs/pull/27) had been opened to add queries as first class objects. This PR did a good job to raise awareness of the need for having queries as first-class objects but this PR expands on that in a few areas:
```

Comment:
I put quite some concerns over https://github.com/amundsen-io/rfcs/pull/27 . Please take a look.

Reply:
let me pull them in so they're in one place 

> typically approach is to store the query log through a data warehouse table (delta, hive etc) and it could fetch the usage log and persist the relevant information into the metadata store.
> I wonder how. the usage query log ingestion is going to look like? are you directly consume from the source and directly persist into metastore store?

This is supported in the updated proposal. Some systems have it builtin, but most installs we've seen use a streaming service to dump the log data into an object store (e.g. Kinesis into S3), then they do batch aggregations into Amundsen. But overall it's been pretty disparate, and I don't feel super confident that we'll be able to build turn-key integrations that will work great for all installs. So we'd like to make all the building blocks available, and leave the actual aggregation and processing up to the implementor so they can fit it into their infrastructure.


> have we done any tests on how long it takes to ingest user log sql. Typically production usage log is 10s of millions of sql per day. Are we planning to surface all the queries in the table page?

The table detail page will only have the top 2-10 queries, configurable. I don't even think it needs to be paginated, a simple leaderboard is sufficient for most of the use cases we've seen.

> if not during ingest, when will the sql parsing happen? are we planning to expose the usagelog for lineage consumption through API?

Can you clarify your question about usage log for lineage?


> could you clarify the audit logging piece? also will the usage log ingestion refreshed everyday?

We're de-pri-ing the audit log use case here, since we're no longer storing every single invocation.


> there is already a https://github.com/amundsen-io/amundsendatabuilder/blob/master/databuilder/models/dashboard/dashboard_query.py, we should at least make a base query class and make two subclasses (dashboard, table)

👍 


> I think having a table query relationship makes sense, however, I am not sure about using amundsen as the source of truth for all the queries which could easily store many non-sense queries(e.g typo etc).

> SLA of query log emit from computation framework to this store is NA (typical is to hive table)
> Non GC of the query nodes, are you going to retain PII data(query has all kinds PII) for the full deployment?
> I think we could have the query model, but if I am going to use it, I will only surface the most relevant / popular queries to the table (top 5, or top X)

Yep, this should all be easy to support with these models. Basically, implementors can GC the query agg nodes after 7-30 days (whatever they want) in the graph, and then all of the orphan nodes can be GC'd.

And because the ingest layer is deciding what to surface (and pre-aggregating), the total amount of data is bound.